File: src/main/java/com/httpain/constraints/Assignment.java
package com.httpain.constraints;

import org.optaplanner.core.api.domain.solution.PlanningEntityCollectionProperty;
import org.optaplanner.core.api.domain.solution.PlanningScore;
import org.optaplanner.core.api.domain.solution.PlanningSolution;
import org.optaplanner.core.api.domain.solution.ProblemFactCollectionProperty;
import org.optaplanner.core.api.domain.valuerange.ValueRangeProvider;
import org.optaplanner.core.api.score.buildin.hardsoft.HardSoftScore;

import java.util.List;

@PlanningSolution
public class Assignment {
    @ValueRangeProvider
    @ProblemFactCollectionProperty
    private List<Character> characters;

    @PlanningEntityCollectionProperty
    private List<Group> groups;

    @PlanningScore
    private HardSoftScore score;

    public Assignment() {
    }

    public Assignment(List<Character> characters, List<Group> groups) {
        this.characters = characters;
        this.groups = groups;
    }

    public List<Character> getCharacters() {
        return characters;
    }

    public List<Group> getGroups() {
        return groups;
    }

    public HardSoftScore getScore() {
        return score;
    }

    public void setScore(HardSoftScore score) {
        this.score = score;
    }
}

File: src/main/kotlin/com/httpain/constraints/Assignment.kt
package com.httpain.constraints

import ai.timefold.solver.core.api.domain.solution.PlanningEntityCollectionProperty
import ai.timefold.solver.core.api.domain.solution.PlanningScore
import ai.timefold.solver.core.api.domain.solution.PlanningSolution
import ai.timefold.solver.core.api.domain.solution.ProblemFactCollectionProperty
import ai.timefold.solver.core.api.domain.valuerange.ValueRangeProvider
import ai.timefold.solver.core.api.score.buildin.hardsoft.HardSoftScore

@PlanningSolution
@NoArgConstructor
data class Assignment(

    @PlanningEntityCollectionProperty
    val characters: List<Character>,

    @ValueRangeProvider
    @ProblemFactCollectionProperty
    val groups: List<Group>,

    @PlanningScore
    var score: HardSoftScore? = null
)